icechunk/src/repository.rs

```diff
@@ -178,34 +186,71 @@ pub enum RepositoryError {
     DeserializationError(#[from] rmp_serde::decode::Error),
     #[error("error finding conflicting path for node `{0}`, this probably indicades a bug in `rebase`")]
     ConflictingPathNotFound(NodeId),
+    #[error("error in config deserialization `{0}`")]
+    ConfigDeserializationError(#[from] serde_yml::Error),
 }
 
 pub type RepositoryResult<T> = Result<T, RepositoryError>;
 
-// FIXME: what do we want to do with implicit groups?
-//
 impl Repository {
-    pub fn update(
+    pub async fn fetch_config(
+        storage: &(dyn Storage + Send + Sync),
+    ) -> RepositoryResult<Option<(RepositoryConfig, ETag)>> {
+        match storage.fetch_config().await? {
+            Some((bytes, etag)) => {
+                let config = serde_yml::from_slice(&bytes)?;
+                Ok(Some((config, etag)))
+            }
+            None => Ok(None),
+        }
+    }
+
+    pub async fn update(
         storage: Arc<dyn Storage + Send + Sync>,
         previous_version_snapshot_id: SnapshotId,
-    ) -> RepositoryBuilder {
-        RepositoryBuilder::new(storage, previous_version_snapshot_id)
+    ) -> RepositoryResult<RepositoryBuilder> {
+        let storage_c1 = Arc::clone(&storage);
+        let storage_c2 = Arc::clone(&storage);
+
+        let snap_id = previous_version_snapshot_id.clone();
+        let h1 = tokio::spawn(async move { storage_c1.fetch_snapshot(&snap_id).await });
+        let h2 =
+            tokio::spawn(async move { Self::fetch_config(storage_c2.as_ref()).await });
+
+        // Fail if the snapshot id cannot be found
+        #[allow(clippy::expect_used)]
+        let _must_exist = h1.await.expect("Thread died unexpectedly")?;
```

Comment:
why not say "snapshot id cannot be found"?

Reply:
this is the unwrap for the `spawn` result, then the `?` is to unwrap the Result. So, we should never see that `expect` trigger. It would fail if somehow the thread executing the task panics